controllers/operatorpolicy_controller.go

```diff
@@ -34,16 +66,137 @@ type OperatorPolicyReconciler struct {
-	return ctrl.Result{}, nil
+	return reconcile.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *OperatorPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&policyv1beta1.OperatorPolicy{}).
-		Complete(r)
+// PeriodicallyExecOperatorPolicies loops through all operatorpolicies in the target namespace and triggers
+// template handling for each one. This function drives all the work the operator policy controller does.
+func (r *OperatorPolicyReconciler) PeriodicallyExecOperatorPolicies(
+	ctx context.Context, freq uint, elected <-chan struct{},
+) {
+	_ = ctx // currently unused
+
+	OpLog.Info("Waiting for leader election before periodically evaluating operator policies")
+	<-elected
+
+	exiting := false
+	for !exiting {
+		start := time.Now()
+
+		var skipLoop bool
+		if !skipLoop {
+			policiesList := policyv1beta1.OperatorPolicyList{}
+
+			err := r.List(context.TODO(), &policiesList)
+			if err != nil {
+				OpLog.Error(err, "Failed to list the OperatorPolicy objects to evaluate")
+			} else {
+				for i := range policiesList.Items {
+					policy := policiesList.Items[i]
+					if !r.shouldEvaluatePolicy(&policy) {
+						continue
+					}
+
+					// handle policy
+					err := r.handleSinglePolicy(&policy)
+					if err != nil {
+						OpLog.Error(err, "Error while evaluating operator policy")
+					}
+				}
+			}
+		}
+
+		elapsed := time.Since(start).Seconds()
+		if float64(freq) > elapsed {
+			remainingSleep := float64(freq) - elapsed
+			sleepTime := time.Duration(remainingSleep) * time.Second
+			OpLog.V(2).Info("Sleeping before reprocessing the operator policies", "seconds", sleepTime)
+			time.Sleep(sleepTime)
+		}
+	}
+}
+
+// handleSinglePolicy encapsulates the logic for processing a single operatorPolicy.
+// Currently, this just means setting the status to Compliant and logging the name
+// of the operatorPolicy.
+//
+// In the future, more reconciliation logic will be added. For reference:
+// https://github.com/JustinKuli/ocm-enhancements/blob/89-operator-policy/enhancements/sig-policy/89-operator-policy-kind/README.md
+func (r *OperatorPolicyReconciler) handleSinglePolicy(
+	policy *policyv1beta1.OperatorPolicy,
+) error {
+	// when testing, evaluate at every interval regardless of policy status
+	if testFlag {
+		policy.Status.ComplianceState = policyv1.Compliant
+
+		err := r.updatePolicyStatus(policy)
+		if err != nil {
+			OpLog.Info("error while updating policy status")
+
+			return err
+		}
+	}
+
+	OpLog.Info("Logging operator policy", "policy", policy.Name)
+
+	return nil
+}
+
+// updatePolicyStatus updates the status of the operatorPolicy.
+//
+// In the future, a condition should be added as well, and this should generate events.
+func (r *OperatorPolicyReconciler) updatePolicyStatus(
+	policy *policyv1beta1.OperatorPolicy,
+) error {
+	updatedStatus := policy.Status
+
+	maxRetries := 3
+	for i := 1; i <= maxRetries; i++ {
```

Comment:
I think we've tried this kind of retry thing before, but it eventually got removed (you've probably been digging through the code more recently than me though, so feel free to correct me). In my opinion, retries here aren't necessary - if it fails, it will just get retried next loop anyway.

Reply:
Especially as the status becomes more complex (when there are more things that could cause errors) I think it will just be easier to return the error immediately, and handle the retry at a higher level.

Reply:
I found where this kind of thing *is* in the normal config-policy-controller... it looks like it was useful there because of some non-idempotent stuff: if the config-policy should create something, but then can't record that it created it in its status, then there will be buggy behavior on a later reconcile because it incorrectly thinks it *didn't* create the resource.

So... let's keep an eye out for that kind of situation in the operator policy controller, and see if this ends up being helpful. I think a better approach to the config-policy situation might have been to add an annotation or label to the object that the config-policy-controller created, that way it could determine later whether it had created it or not even if the status didn't get updated correctly.

Reply:
Thanks for pointing this out, I also believe that it would get retried the next loop. I'll keep an eye out for the situations you described in your last comment.